### PR TITLE
Docs: How JSTZ works? in CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by alanmarkoTrilitech**
# Context

https://app.asana.com/0/1205170694555918/1205457666696791

# Description
This PR adds overview to the CONTRIBUTING.md guide.

### CODE REVIEW

File: docs/how_jstz_works.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+---
+title: 👨‍⚖️ How jstz works?
+---
+
+<script setup>
+import VPButton from "vitepress/dist/client/theme-default/components/VPButton.vue";
+</script>
+
```
Comment (johnyob) [inline docs/how_jstz_works.md:8]: Following our discussion yesterday, this doc will be moving to our internal docs (likely a section in `CONTRIBUTING.md`). So these vitepress headers / buttons are no longer required

File: docs/how_jstz_works.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# 👨‍⚖️ `How jstz works?`
```
Comment (johnyob) [inline docs/how_jstz_works.md:9]: How `jstz` works?

File: docs/how_jstz_works.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## Overview
+
+`jstz` is a JavaScript/Typescript server runtime for Tezos 2.0 designed to provide a great developer experience by aiming to be compatible with web conventions.
```
Comment (johnyob) [inline docs/how_jstz_works.md:13]: JavaScript server runtime. 
Conventions -> standards (https://developer.mozilla.org/en-US/docs/Web/API)

File: docs/how_jstz_works.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+Through `jstz` developers can set up, deploy and test so called _smart functions_ written in Javascript/Typescript that can get directly executed on Tezos L2.
```
Comment (johnyob) [inline docs/how_jstz_works.md:15]: executed on Tezos L2 -> executed on the `jstz` smart rollup node. 
Link https://tezos.gitlab.io/alpha/smart_rollups.html on smart rollup

File: docs/how_jstz_works.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+Jstz provides a local sandboxed environment for developers to test the functions without deploying them to L2. It provides a simple interface through which one can deploy smart functions and then call them by sending HTTP requests to a particular _smart function address_.
```
Comment (johnyob) [inline docs/how_jstz_works.md:17]: Jstz->`jstz`. 
Some grammar changes.
Moved running of smart functions to above paragraph

```suggestion
`jstz` also provides a local sandboxed environment for developers to test their functions without deploying them to production.
```

File: docs/how_jstz_works.md (hunk 4 above)
Comment (johnyob) [reply docs/how_jstz_works.md:15]: Please move the running of smart functions using HTTP requests to this paragraph.

File: docs/how_jstz_works.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+## How it works?
+
+Jstz has to ensure that smart functions written in JS/TS can compile directly to WASM - the assembly used for writing Smart Rollups. Therefore 'jstz' is built on _Boa_ - a Javascript engine written in Rust.
```
Comment (johnyob) [inline docs/how_jstz_works.md:21]: Not entirely correct. smart functions don't have to compile to WASM. smart rollups must compile to WASM, hence `jstz` needs to use a JavaScript engine that compiles to WASM.

File: docs/how_jstz_works.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+In its core (jstz_core), Jstz uses Boa and enables Rust types to be passed around as JavaScript objects. This allows implementation and registration of various APIs written in Rust and their usage as if they were native Javascript objects.
```
Comment (johnyob) [inline docs/how_jstz_works.md:23]: In the `jstz_core` crates, ...

File: docs/how_jstz_works.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+The APIs (implemented in jstz_api) enable interaction with the storage of a rollup kernel (a key-value store) and also implement Request and Response APIs based on the Fetch API specification enabling calling of other smart functions.
+
```
Comment (johnyob) [inline docs/how_jstz_works.md:26]:
```suggestion
`jstz` implements several `jstz`-specific APIs such as `Kv`, `Ledger`, and `Contract`. Additional (in the `jstz_api` crate), `jstz` provides implementations for many web standard APIs. 
```

File: docs/how_jstz_works.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+##
```
Comment (johnyob) [inline docs/how_jstz_works.md:27]:
```suggestion
## `jstz`-specific APIs
```

File: docs/how_jstz_works.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+### KV store
+
+When writing smart functions, we need a way to store data across different calls of the functions. Therefore, jstz _smart functions_ implement a persistent key-value store used for storing and retrieval of arbitrary JSON blobs. This store can be accessed through a global _Kv_ object.
+
+The key-value store implements _optimistic concurrency control scheme_. It is optimistically assumed that conflicts between different transactions (snapshots of the persistent kv store) are sufficiently rare thus not interfering each other. Before commiting, the transaction verifies whether no other transaction has modified the data it has read.
+
+The transactions performed over the KV store offer ACID guarantees and serializability, therefore any transaction can be commited only if it does not conflict with any previously commited ones.
+
+In each transaction, the repeated access to the same key is optimized through caching. Similarly, writes are buffered until the transaction is commited at which point it gets flushed to the persistent KV storage.
```
Comment (johnyob) [inline docs/how_jstz_works.md:37]: Might be good to move these paragraphs into the paragraph mentioning `jstz_core`. And then mention that the `Kv` store is implemented on top of `jstz_core::kv`.

File: docs/how_jstz_works.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+### Ledger
+
+A specialised type of the KV store is the Ledger that provides access to the balances of the L2 tez. Additionally it also stores so-called 'self address' - the address of the smart function itself. Similarly to the KV store, all operations on the ledger are synchronous and atomic, commited only if the request to the smart function succeeds.
+
+### Console/URL/Textencoder
+
+Additionally, jstz implements APIs for easier manipulation with debug logs (Console), with unique smart function addresses (URL) or simplification of encoding of passed JSON parameters (TextEncoder).
+
+##
```
Comment (johnyob) [inline docs/how_jstz_works.md:47]: Don't think we need to explain web standard APIs

File: docs/how_jstz_works.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+- the APIs get registered to in the Realm that consists of a set of intrinsic objects and global environment
+- The Realm wrapper implements various methods for registration and evaluation of different modules, types and host defined objects and handling of context
+- JSNative permits Rust types to be passed around as JavaScript objects.
+- There is implemented a wrapper over boa engines runtime and also a wrapper over the smart rollup's runtime - erased runtime.
+- the APIs use the functionality of the rollup runtime to interact with the blockchain storage and other functionality implemented in jstz_proto
+- jstz_kernel
+
+## Bridge
+
+In order to transfer tez from L1 to an L2 address used by a smart function running on a `jstz` rollup, `jstz` implements a simple ticket-based jstz bridge smart contracts, through which it is possible to deposit tez from an L1 address to a jstz rollup address.
```
Comment (johnyob) [inline docs/how_jstz_works.md:60]:
```suggestion
In order to transfer ctez from L1 address to an L2 address in `jstz`, `jstz` implements a simple ticket-based bridge smart contract built with LIGO. This contract enables users to deposit ctez from an L1 address (`tz1`/`KT1`) to a jstz address (`tz4`).
```

Comment (johnyob) [reply docs/how_jstz_works.md:60]: Would be also good to get an example of depositing

File: docs/how_jstz_works.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+##
+
+//TODO: Explaining how exactly the following works and fits together:
+
+- the APIs get registered to in the Realm that consists of a set of intrinsic objects and global environment
+- The Realm wrapper implements various methods for registration and evaluation of different modules, types and host defined objects and handling of context
+- JSNative permits Rust types to be passed around as JavaScript objects.
+- There is implemented a wrapper over boa engines runtime and also a wrapper over the smart rollup's runtime - erased runtime.
+- the APIs use the functionality of the rollup runtime to interact with the blockchain storage and other functionality implemented in jstz_proto
+- jstz_kernel
```
Comment (johnyob) [inline docs/how_jstz_works.md:56]: Comment these out. We will address these points in a follow up PR

File: docs/how_jstz_works.md (hunk 12 above)
Comment (alanmarkoTrilitech) [reply docs/how_jstz_works.md:60]: Not sure what you by an example - If just how running of the command `jstz bridge deposit --from ... --to --amount` - I don't think it fits here as it is shown in the CLI doc.

**Comment by johnyob** [comment]
LGTM, would it be possible to move the doc contents to `CONTRIBUTING.md` before merging

**Review by johnyob — APPROVED** [review]

**Pull request merged**